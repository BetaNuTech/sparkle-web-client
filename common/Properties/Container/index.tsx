import { ReactElement, FunctionComponent } from 'react';
import styles from './styles.module.scss';

interface Props {
  header: ReactElement;
  grid: ReactElement;
  sidebar?: ReactElement;
}

const Container: FunctionComponent<Props> = ({ header, grid, sidebar }) => (
  <div className={styles.container}>
    {header}

    <div className={styles.container__main}>{grid}</div>
    <aside>{sidebar}</aside>
  </div>
);

export default Container;
