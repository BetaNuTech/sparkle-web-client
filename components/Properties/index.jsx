import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { calculateTeamValues } from '../../utils/calculateTeamValues';
import {
  activePropertiesSortFilter,
  sortProperties
} from '../../utils/propertiesSorting';
import {
  fetchDataOfProperties,
  setDataOfProperties
} from '../../redux/ducks/properties/actionCreators';
import {
  selectActiveSortOfProperties,
  selectItemsOfProperties
} from '../../redux/ducks/properties/selectors';
import { fetchDataOfTeams } from '../../redux/ducks/teams/actionCreators';
import { selectItemsOfTeams } from '../../redux/ducks/teams/selectors';
import styles from './Properties.module.scss';
import { Header } from './Header';
import { Sidebar } from './Sidebar';
import { ProfileList } from './ProfileList';
import { MobileProperties } from './MobileProperties';

export const Properties = () => {
  const dispatch = useDispatch();
  const [teamCalculatedValues, setTeamCalculatedValues] = useState([]);
  const activeSort = useSelector(selectActiveSortOfProperties);
  const properties = useSelector(selectItemsOfProperties);
  const teams = useSelector(selectItemsOfTeams);

  // We fetch data only if store with properties is empty.
  // When we switch the page with full store, it doesn't fetch.
  useEffect(() => {
    dispatch(fetchDataOfTeams());

    if (properties.length === 0) {
      dispatch(fetchDataOfProperties());
    }
  }, []);

  // We recalculate properties when properties or teams changes.
  useEffect(() => {
    setTeamCalculatedValues(calculateTeamValues(teams, properties));
  }, [properties, teams]);

  // We sort  properties when activeSort changes.
  useEffect(() => {
    dispatch(
      setDataOfProperties(
        [...properties].sort(
          sortProperties(activeSort.sortBy, activeSort.orderBy)
        )
      )
    );
  }, [activeSort]);

  return (
    <>
      <div className={styles.properties__sortInfoLine}>
        {`Sorted by ${activePropertiesSortFilter(activeSort.sortBy)}`}
      </div>

      <div className={styles.properties__wrapper}>
        <header>
          <Header activeSort={activeSort} properties={properties} />
        </header>

        <div className={styles.properties__main}>
          <ProfileList activeSort={activeSort} properties={properties} />
        </div>

        <aside>
          <Sidebar teams={teams} teamCalculatedValues={teamCalculatedValues} />
        </aside>
      </div>

      <div className={styles.properties__mobile}>
        <MobileProperties
          properties={properties}
          teams={teams}
          teamCalculatedValues={teamCalculatedValues}
        />
      </div>
    </>
  );
};
