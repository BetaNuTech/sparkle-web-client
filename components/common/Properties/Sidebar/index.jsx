import styles from './Sidebar.module.scss';
import { Item } from './Item';

export const Sidebar = () => (
  <nav className={styles.sidebar}>
    <h4 className={styles.sidebar__heading}>Teams</h4>

    <ul className={styles.sidebar__list}>
      <Item name="Team one" />
      <Item name="Team two" />
      <Item name="Team three" />
    </ul>
  </nav>
);
