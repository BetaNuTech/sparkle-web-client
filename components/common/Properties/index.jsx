import React, { useState } from 'react';
import styles from './Properties.module.scss';
import { Header } from './Header';
import { ProfileList } from './ProfileList';
import { Sidebar } from './Sidebar';
import { MobileProperties } from './MobileProperties';

export const Properties = () => {
  const [isAscendingSort, setIsAscendingSort] = useState(true);

  return (
    <div className={styles.properties}>
      <div className={styles.properties__wrapper}>
        <header>
          <Header
            isAscendingSort={isAscendingSort}
            setIsAscendingSort={setIsAscendingSort}
          />
        </header>

        <div className={styles.properties__main}>
          <ProfileList isAscendingSort={isAscendingSort} />
        </div>

        <aside>
          <Sidebar />
        </aside>
      </div>

      <div className={styles.properties__mobile}>
        <MobileProperties />
      </div>
    </div>
  );
};
