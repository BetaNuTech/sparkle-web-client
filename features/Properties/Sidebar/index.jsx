import PropTypes from 'prop-types';
import styles from './Sidebar.module.scss';
import { TeamItem } from './TeamItem';

export const Sidebar = ({ teams, teamCalculatedValues }) => (
  <nav className={styles.sidebar} data-testid="properties-teams-sidebar">
    <h4 className={styles.sidebar__heading}>Teams</h4>

    <ul className={styles.sidebar__list}>
      {teamCalculatedValues.length !== 0 &&
        teams.map((team) => (
          <TeamItem
            key={team.id}
            team={team}
            teamCalculatedValues={teamCalculatedValues[team.id - 1]}
          />
        ))}
    </ul>
  </nav>
);

Sidebar.propTypes = {
  teams: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      properties: PropTypes.arrayOf(PropTypes.number).isRequired,
      propertyCount: PropTypes.number.isRequired
    }).isRequired
  ).isRequired,
  teamCalculatedValues: PropTypes.arrayOf(
    PropTypes.shape({
      totalNumOfDeficientItems: PropTypes.number.isRequired,
      totalNumOfFollowUpActionsForDeficientItems: PropTypes.number.isRequired,
      totalNumOfRequiredActionsForDeficientItems: PropTypes.number.isRequired
    }).isRequired
  ).isRequired
};
