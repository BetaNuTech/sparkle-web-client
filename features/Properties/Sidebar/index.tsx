import { FunctionComponent } from 'react';
import styles from './styles.module.scss';
import TeamItem from './TeamItem';
import propertyMetaData from '../../../common/models/propertyMetaData';

interface PropertiesSidebarTeamItemWrapperModel {
  team: any;
  teamCalculatedValues: Array<propertyMetaData>;
}

interface PropertiesSidebarLayoutModel {
  teams: Array<any>;
  teamCalculatedValues: Array<propertyMetaData>;
}

// Wrapper around team items
// to lookup properties meta data
const TeamItemWrapper: FunctionComponent<PropertiesSidebarTeamItemWrapperModel> =
  ({ team, teamCalculatedValues }) => {
    const [propertiesMetaData] = teamCalculatedValues.filter(
      ({ team: teamId }) => teamId === team.id
    );

    return <TeamItem team={team} teamCalculatedValues={propertiesMetaData} />;
  };

// Sidebar component
const PropertiesSidebar: FunctionComponent<PropertiesSidebarLayoutModel> = ({
  teams,
  teamCalculatedValues
}) => (
  <nav className={styles.sidebar} data-testid="properties-teams-sidebar">
    <h4 className={styles.sidebar__heading}>Teams</h4>

    <ul className={styles.sidebar__list}>
      {teamCalculatedValues.length &&
        teams.map((team) => (
          <TeamItemWrapper
            key={team.id}
            team={team}
            teamCalculatedValues={teamCalculatedValues}
          />
        ))}
    </ul>
  </nav>
);

export default PropertiesSidebar;
