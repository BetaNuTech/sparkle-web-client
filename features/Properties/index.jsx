import PropTypes from 'prop-types';
import { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { calculateTeamValues } from '../../common/utils/calculateTeamValues';
import {
  sorts,
  activePropertiesSortFilter,
  sortProperties
} from '../../common/utils/propertiesSorting';
import {
  fetchDataOfProperties,
  setDataOfProperties,
  setActiveSortOfProperties
} from '../../app/ducks/properties/actionCreators';
import {
  selectActiveSortOfProperties,
  selectItemsOfProperties
} from '../../app/ducks/properties/selectors';
import { fetchDataOfTeams } from '../../app/ducks/teams/actionCreators';
import { selectItemsOfTeams } from '../../app/ducks/teams/selectors';
import styles from './Properties.module.scss';
import { Header } from './Header';
import { Sidebar } from './Sidebar';
import { ProfileList } from './ProfileList';
import { MobileHeader } from './MobileHeader';
import { MobileProperties } from './MobileProperties';

export const Properties = ({
  isOnline,
  isStaging,
  isNavOpen,
  toggleNavOpen
}) => {
  const dispatch = useDispatch();
  const [teamCalculatedValues, setTeamCalculatedValues] = useState([]);
  const activeSort = useSelector(selectActiveSortOfProperties);
  const activeSortDir = { ...activeSort };
  const properties = useSelector(selectItemsOfProperties);
  const teams = useSelector(selectItemsOfTeams);

  // Fetch data only if store with properties is empty.
  // When we switch the page with full store, it doesn't fetch.
  useEffect(() => {
    dispatch(fetchDataOfTeams());

    if (properties.length === 0) {
      dispatch(fetchDataOfProperties());
    }
  }, []);

  // Recalculate properties when properties or teams changes.
  useEffect(() => {
    setTeamCalculatedValues(calculateTeamValues(teams, properties));
  }, [properties, teams]);

  // Sort properties active sort and direction
  useEffect(() => {
    dispatch(
      setDataOfProperties(
        [...properties].sort(
          sortProperties(activeSort.sortBy, activeSort.orderBy)
        )
      )
    );
  }, [activeSort]);

  // Loop through property
  // sorting options
  const nextPropertiesSort = () => {
    const activeFilter =
      sorts[sorts.indexOf(activeSort.sortBy) + 1] || sorts[0]; // Get next or first

    // Update Property Filter
    dispatch(
      setActiveSortOfProperties({
        ...activeSort,
        sortBy: activeFilter
      })
    );
  };

  // Set sort attribute & direction
  const onSortChange = (key) => (event) => {
    const {
      target: { value }
    } = event;

    if (key === 'orderBy') {
      // Update orderBy filter
      if (activeSortDir[key] === 'asc') activeSortDir[key] = 'desc';
      else if (activeSortDir[key] === 'desc') activeSortDir[key] = 'asc';
    } else {
      // Update sortBy filter
      activeSortDir[key] = value;
    }

    // Update Property Filter
    dispatch(setActiveSortOfProperties(activeSortDir));
  };

  return (
    <>
      <MobileHeader
        title="Properties"
        isNavOpen={isNavOpen}
        toggleNavOpen={toggleNavOpen}
        nextPropertiesSort={nextPropertiesSort}
        isOnline={isOnline}
        isStaging={isStaging}
      />

      <div className={styles.properties__sortInfoLine}>
        {`Sorted by ${activePropertiesSortFilter(activeSort.sortBy)}`}
      </div>

      <div className={styles.properties__wrapper}>
        <Header activeSort={activeSort} onSortChange={onSortChange} />

        <div className={styles.properties__main}>
          <ProfileList activeSort={activeSort} properties={properties} />
        </div>

        <aside>
          <Sidebar teams={teams} teamCalculatedValues={teamCalculatedValues} />
        </aside>
      </div>

      <div className={styles.properties__mobile}>
        <MobileProperties
          properties={properties}
          teams={teams}
          teamCalculatedValues={teamCalculatedValues}
        />
      </div>
    </>
  );
};

Properties.propTypes = {
  isOnline: PropTypes.bool,
  isStaging: PropTypes.bool,
  isNavOpen: PropTypes.bool,
  toggleNavOpen: PropTypes.func
};

Properties.defaultProps = {
  isOnline: false,
  isStaging: false,
  isNavOpen: false,
  toggleNavOpen: () => {}
};
