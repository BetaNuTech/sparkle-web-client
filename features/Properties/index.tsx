import { FunctionComponent, useState, useEffect } from 'react';
import { useMediaQuery } from 'react-responsive';
import { useFirestore } from 'reactfire';
import calculateTeamValues from './utils/calculateTeamValues';
import {
  sorts,
  activePropertiesSortFilter,
  sortProperties
} from './utils/propertiesSorting';
import { useSortBy, useSortDir } from './hooks/sorting';
import useTeams from './hooks/useTeams';
import useProperties from './hooks/useProperties';
import useDeleteProperty from '../../common/hooks/useDeleteProperty';
import useDeleteTeam from './hooks/useDeleteTeam';
import {
  canCreateTeam,
  canCreateProperty
} from '../../common/utils/userPermissions';
import useNotifications from '../../common/hooks/useNotifications'; // eslint-disable-line
import userModel from '../../common/models/user';
import propertyModel from '../../common/models/property';
import teamModel from '../../common/models/team';
import notifications from '../../common/services/notifications';
import globalEvents from '../../common/utils/globalEvents';
import breakpoints from '../../config/breakpoints';
import styles from './styles.module.scss';
import Header from './Header';
import Sidebar from './Sidebar';
import ProfileList from './ProfileList';
import MobileLayout from './MobileLayout';
import DeleteTeamPrompt from './DeleteTeamPrompt';

interface PropertiesModel {
  user: userModel;
  isOnline?: boolean;
  isStaging?: boolean;
  isNavOpen?: boolean;
  toggleNavOpen?(): void;
  forceVisible?: boolean;
}

const Properties: FunctionComponent<PropertiesModel> = ({
  user,
  isOnline,
  isStaging,
  toggleNavOpen,
  forceVisible
}) => {
  const firestore = useFirestore();
  const [teamCalculatedValues, setTeamCalculatedValues] = useState([]);
  const [sortBy, setSortBy] = useSortBy();
  const [sortDir, setSortDir] = useSortDir();
  const { data: properties, memo: propertiesMemo } = useProperties(
    firestore,
    user
  );
  const {
    status: teamsStatus,
    data: teams,
    memo: teamsMemo
  } = useTeams(firestore, user);
  const [sortedProperties, setSortedProperties] = useState([]);

  // Lookup user permissions
  const hasCreateTeamPermission = canCreateTeam(user);
  const hasCreatePropertyPermission = canCreateProperty(user);

  // User notifications setup
  /* eslint-disable */
  const sendNotification = notifications.createPublisher(useNotifications());
  /* eslint-enable */

  // Queue and Delete Property
  const [isDeletePropertyPromptVisible, setDeletePropertyPromptVisible] =
    useState(false);
  const { queuePropertyForDelete, confirmPropertyDelete } = useDeleteProperty(
    firestore,
    sendNotification,
    user
  );
  const openPropertyDeletePrompt = (property: propertyModel) => {
    queuePropertyForDelete(property);
    setDeletePropertyPromptVisible(true);
  };
  const closeDeletePropertyPrompt = () => {
    setDeletePropertyPromptVisible(false);
    queuePropertyForDelete(null);
  };

  // Queue and Delete Team
  const [isDeleteTeamPromptVisible, setDeleteTeamPromptVisible] =
    useState(false);
  const { queueTeamForDelete, confirmTeamDelete } = useDeleteTeam(
    firestore,
    sendNotification,
    user
  );
  const openTeamDeletePrompt = (team: teamModel) => {
    queueTeamForDelete(team);
    setDeleteTeamPromptVisible(true);
  };
  const closeDeleteTeamPrompt = () => {
    setDeleteTeamPromptVisible(false);
    queueTeamForDelete(null);
  };

  // Responsive queries
  const isMobileorTablet = useMediaQuery({
    maxWidth: breakpoints.tablet.maxWidth
  });
  const isDesktop = useMediaQuery({
    minWidth: breakpoints.desktop.minWidth
  });

  // Apply properties sort order
  const applyPropertiesSort = () =>
    [...properties].sort(sortProperties(sortBy, sortDir));

  // Update team calculated values
  useEffect(() => {
    function recalcTeamComputedValues() {
      if (teamsStatus === 'success') {
        setTeamCalculatedValues(calculateTeamValues(teams, properties));
      }
    }

    recalcTeamComputedValues();
  }, [teamsStatus, teamsMemo, propertiesMemo]); // eslint-disable-line

  useEffect(() => {
    function resortProperties() {
      setSortedProperties(applyPropertiesSort());
    }

    resortProperties();
  }, [propertiesMemo, sortBy, sortDir]); // eslint-disable-line

  // Loop through property
  // sorting options
  const nextPropertiesSort = () => {
    const activeSortValue = sorts[sorts.indexOf(sortBy) + 1] || sorts[0]; // Get next or first
    const descOrderKeys = ['lastInspectionDate', 'lastInspectionScore'];
    const isDescOrderValue = descOrderKeys.includes(activeSortValue);

    // Check if sort direction is asc then
    // change to desc if it is either lastInspectionDate or lastInspectionScore
    if (isDescOrderValue && sortDir === 'asc') {
      setSortDir('desc');
    } else if (!isDescOrderValue && sortDir === 'desc') {
      // If sort desc and it is other than values
      // in `descOrderKeys` then change to asc
      setSortDir('asc');
    }
    // Update Property sort
    setSortBy(activeSortValue);
    globalEvents.trigger('visibilityForceCheck');
  };

  // Set sort attribute & direction
  const onSortChange = (key: string) => (evt: { target: HTMLInputElement }) => {
    const {
      target: { value }
    } = evt;

    // Update sort direction
    if (key === 'sortDir') {
      setSortDir(sortDir === 'asc' ? 'desc' : 'asc');
    } else {
      setSortBy(value); // Update sort by
    }

    globalEvents.trigger('visibilityForceCheck');
  };

  return (
    <>
      {isMobileorTablet && (
        <MobileLayout
          properties={sortedProperties}
          teams={teams}
          teamCalculatedValues={teamCalculatedValues}
          isDeletePropertyPromptVisible={isDeletePropertyPromptVisible}
          confirmPropertyDelete={confirmPropertyDelete}
          openPropertyDeletePrompt={openPropertyDeletePrompt}
          closeDeletePropertyPrompt={closeDeletePropertyPrompt}
          openTeamDeletePrompt={openTeamDeletePrompt}
          isOnline={isOnline}
          isStaging={isStaging}
          canAddTeam={hasCreateTeamPermission}
          canAddProperty={hasCreatePropertyPermission}
          toggleNavOpen={toggleNavOpen}
          nextPropertiesSort={nextPropertiesSort}
          sortBy={sortBy}
          activePropertiesSortFilter={activePropertiesSortFilter}
          forceVisible={forceVisible}
        />
      )}

      {/* Desktop Header & Content */}
      {isDesktop && (
        <div className={styles.properties__container}>
          <Header
            sortBy={sortBy}
            sortDir={sortDir}
            onSortChange={onSortChange}
            canAddTeam={hasCreateTeamPermission}
            canAddProperty={hasCreatePropertyPermission}
          />

          <div className={styles.properties__main}>
            <ProfileList
              properties={sortedProperties}
              forceVisible={forceVisible}
            />
          </div>

          <aside>
            <Sidebar
              teams={teams}
              openTeamDeletePrompt={openTeamDeletePrompt}
              teamCalculatedValues={teamCalculatedValues}
            />
          </aside>
        </div>
      )}

      <DeleteTeamPrompt
        isVisible={isDeleteTeamPromptVisible}
        onClose={closeDeleteTeamPrompt}
        onConfirm={confirmTeamDelete}
      />
    </>
  );
};

Properties.defaultProps = {
  isOnline: false,
  isStaging: false,
  isNavOpen: false,
  toggleNavOpen: () => {}, // eslint-disable-line
  forceVisible: false
};

export default Properties;
