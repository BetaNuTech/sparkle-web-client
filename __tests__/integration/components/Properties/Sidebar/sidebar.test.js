import React from 'react';
import { shallow } from 'enzyme';
import { Sidebar } from '../../../../../components/Properties/Sidebar';

describe('Sidebar component', () => {
  describe('rendering', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<Sidebar />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
