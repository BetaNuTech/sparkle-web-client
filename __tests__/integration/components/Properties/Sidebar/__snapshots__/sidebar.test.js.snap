// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component rendering renders correctly 1`] = `ShallowWrapper {}`;
